[Sanity_Test_Failing] The ICS timer expiry handling after mme restart is failing

### Your Environment

- **Version:** Local Development Setup
- **Affected Component:** Access Gateway
- **Affected Subcomponent:** Magma MME/SPGW service
- **Deployment Environment:** Vagrant (AGW)

### Setup Details
The S1APTester ([https://github.com/facebookexperimental/S1APTester](https://github.com/facebookexperimental/S1APTester)) was used along with Magma EPC in Vagrant based setup to reproduce the issue.

### Describe the Issue
The sanity test case **s1aptests/test_ics_timer_expiry_with_mme_restart.py** is always getting stuck waiting for response from MME and fails with timeout. The test case expects the UE context release after attach but does not receive the response and gets stuck. 

On the MME side, as per the current timer implementation, the timer id is created when starting any timer only and based on the timer id mme_ue_s1ap_id is retrieved on any timer expiry and then handled accordingly. The ICS timer expiry is of 4 seconds. For the above mentioned test case, the timer has already expired by the time MME comes up after restart. But since the timer has already expired, it does not resume timer to create new timer id and rather calls for timer expiry handling. The timer expiry handling fails to fetch mme_ue_s1ap_id based on timer id, because timer id is an invalid one since we don't restore timer ids from redis db after mme restart.

Proposed solution: Need to store timer id and mme_ue_s1ap_id mapping to redis db before MME restart and restore them before timer expiry handling.

Point of culprit:
```
Function mme_app_recover_timers_for_ue in file magma/lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.c
Function mme_app_handle_initial_context_setup_rsp_timer_expiry in file magma/lte/gateway/c/core/oai/tasks/mme_app/mme_app_bearer.c
```

The error log in the MME are:
```
000399 Wed Mar 02 07:46:31 2022 7FA00D718700 DEBUG MME-AP tasks/mme_app/mme_app_timer_mana:0056         Handling :Initial Context Setup Response timer
000400 Wed Mar 02 07:46:31 2022 7FA00D718700 ERROR MME-AP tasks/mme_app/mme_app_bearer.c  :2042            Invalid Timer Id expiration, Timer Id: 4294967295

```

The terminal error log is as follows:
```
(python) vagrant@magma-test:~/magma/lte/gateway/python/integ_tests$ make integ_test TESTS=s1aptests/test_ics_timer_expiry_with_mme_restart.py
. /home/vagrant/build/python/bin/activate
echo "Running test: s1aptests/test_ics_timer_expiry_with_mme_restart.py"
Running test: s1aptests/test_ics_timer_expiry_with_mme_restart.py
timeout --foreground -k 930s 900s sudo -E PATH=/home/vagrant/build/python/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/go/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-flaky --force-flaky --no-success-flaky-report --max-runs=1 --min-passes=1 --with-xunit --xunit-file=/var/tmp/test_results/test_ics_timer_expiry_with_mme_restart.xml -x -s s1aptests/test_ics_timer_expiry_with_mme_restart.py || exit 1
Start time 07:46:08
tag: AUTH_TYPE tagVal: MILENAGE
************************* Enb tester config

**********************TPT open server******** 0

 Server Socket FD =[13]
************* Testing *Time elapsed(in usec): for timer expiry               :9263 usec
AGW is stateless
APN Correction configured
Health service is disabled
Using subscriber IMSI IMSI001010000000001
Using IMEI 3805468432113171
************************* UE device config for ue_id  1
************************* Configuring IP block
************************* Waiting for IP changes to propagate
************************* S1 setup
************************* UE App config
************************* Running End to End attach for UE id  1
ue id 1 found

*************NbUiNbuHdlUeRadioCapMsg: Send UE cap msg SUCCESS*********
PDN_TYPE: 1
IPv4 PDN_ADDRESS:192 168 128 12
************************* Sending UE context release request  for UE id  1
Deleting Ue Context from Traffic Handler
*** Sending indication to drop Initial Context Setup Req ***
************************* Sending Service request for UE id  1
************************* Restarting MME service on gateway
Waiting for 0 seconds
Waiting for 1 seconds
Waiting for 2 seconds
Waiting for 3 seconds
Waiting for 4 seconds
Waiting for 5 seconds
Waiting for 6 seconds
Waiting for 7 seconds
Waiting for 8 seconds
Waiting for 9 seconds
Waiting for 10 seconds
Waiting for 11 seconds
Waiting for 12 seconds
Waiting for 13 seconds
Waiting for 14 seconds
Waiting for 15 seconds
Waiting for 16 seconds
Waiting for 17 seconds
Waiting for 18 seconds
Waiting for 19 seconds
Waiting for 20 seconds
Waiting for 21 seconds
Waiting for 22 seconds
Waiting for 23 seconds
Waiting for 24 seconds
Waiting for 25 seconds
Waiting for 26 seconds
Waiting for 27 seconds
Waiting for 28 seconds
Waiting for 29 seconds
************************* send SCTP SHUTDOWN
Keys left in Redis (list should be empty)[
 IMSI001010000000001.magma.ipv4,ipv4:mobilityd_ipdesc_record
IMSI001010000000001:SPGW
IMSI001010000000001:MME

nb_ue_attached: 1
]
Entries in s1ap_imsi_map (should be zero): 1
Entries left in hashtables (should be zero): 3
Entries in mme_ueip_imsi_map (should be zero): 1
F
======================================================================
FAIL: The test case validates the functionality of ICS timer expiry while UE
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/test_ics_timer_expiry_with_mme_restart.py", line 118, in test_ics_timer_expiry_with_mme_restart
    response = self._s1ap_wrapper.s1_util.get_response()
  File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py", line 212, in get_response
    + " sec) occurred while waiting for response message",
AssertionError: Timeout (900 sec) occurred while waiting for response message
-------------------- >> begin captured logging << --------------------
root: ERROR: Could not load magmad yml config for mconfig modules
urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): 192.168.60.142:3333
urllib3.connectionpool: DEBUG: http://192.168.60.142:3333 "GET /stats/switches HTTP/1.1" 200 17
urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): 192.168.60.142:3333
urllib3.connectionpool: DEBUG: http://192.168.60.142:3333 "GET /stats/switches HTTP/1.1" 200 17
root: INFO: Deleting subscriber : IMSI001010000000001
root: INFO: Adding subscriber : IMSI001010000000001
root: INFO: Ignoring FAILED_PRECONDITION exception
root: DEBUG: s1ap response expected, received: 12, 12
root: DEBUG: s1ap message expected, received: 68, 68
root: INFO: Deleting subscriber : IMSI001010000000001
--------------------- >> end captured logging << ---------------------

Ran 1 test in 957.139s

FAILED (failures=1)
Makefile:44: recipe for target 'integ_test' failed
make: *** [integ_test] Error 1
(python) vagrant@magma-test:~/magma/lte/gateway/python/integ_tests$
```


**To Reproduce**
1) Run the test case in loop with "make integ_test TESTS=s1aptests/test_ics_timer_expiry_with_mme_restart.py"


CC: @rsarwad

--- a/lte/gateway/c/core/oai/tasks/mme_app/mme_app_bearer.c
+++ b/lte/gateway/c/core/oai/tasks/mme_app/mme_app_bearer.c
@@ -1951,13 +1951,19 @@ status_code_e mme_app_handle_mobile_reachability_timer_expiry(zloop_t* loop,
   OAILOG_FUNC_IN(LOG_MME_APP);
 
   mme_ue_s1ap_id_t mme_ue_s1ap_id = 0;
-  if (!mme_pop_timer_arg_ue_id(timer_id, &mme_ue_s1ap_id)) {
-    OAILOG_WARNING(LOG_MME_APP, "Invalid Timer Id expiration, Timer Id: %u\n",
-                   timer_id);
-    OAILOG_FUNC_RETURN(LOG_MME_APP, RETURNok);
+  struct ue_mm_context_s* ue_context_p = NULL;
+  if (timer_id == MME_APP_TIMER_INACTIVE_ID) {
+    // Expiry handler was called as part of timer recovery on service restart
+    ue_context_p = (struct ue_mm_context_s*)args;
+  } else {
+    if (!mme_pop_timer_arg_ue_id(timer_id, &mme_ue_s1ap_id)) {
+      OAILOG_WARNING(LOG_MME_APP, "Invalid Timer Id expiration, Timer Id: %u\n",
+                     timer_id);
+      OAILOG_FUNC_RETURN(LOG_MME_APP, RETURNok);
+    }
+    ue_context_p = mme_app_get_ue_context_for_timer(
+        mme_ue_s1ap_id, "Mobile reachability timer");
   }
-  struct ue_mm_context_s* ue_context_p = mme_app_get_ue_context_for_timer(
-      mme_ue_s1ap_id, "Mobile reachability timer");
   if (ue_context_p == NULL) {
     OAILOG_ERROR(
         LOG_MME_APP,
@@ -2008,13 +2014,20 @@ status_code_e mme_app_handle_implicit_detach_timer_expiry(zloop_t* loop,
                                                           void* args) {
   OAILOG_FUNC_IN(LOG_MME_APP);
   mme_ue_s1ap_id_t mme_ue_s1ap_id = 0;
-  if (!mme_pop_timer_arg_ue_id(timer_id, &mme_ue_s1ap_id)) {
-    OAILOG_WARNING(LOG_MME_APP, "Invalid Timer Id expiration, Timer Id: %u\n",
-                   timer_id);
-    OAILOG_FUNC_RETURN(LOG_MME_APP, RETURNok);
+  struct ue_mm_context_s* ue_context_p = NULL;
+  if (timer_id == MME_APP_TIMER_INACTIVE_ID) {
+    // Expiry handler was called as part of timer recovery on service restart
+    ue_context_p = (struct ue_mm_context_s*)args;
+  } else {
+    if (!mme_pop_timer_arg_ue_id(timer_id, &mme_ue_s1ap_id)) {
+      OAILOG_WARNING(LOG_MME_APP, "Invalid Timer Id expiration, Timer Id: %u\n",
+                     timer_id);
+      OAILOG_FUNC_RETURN(LOG_MME_APP, RETURNok);
+    }
+    ue_context_p = mme_app_get_ue_context_for_timer(mme_ue_s1ap_id,
+                                                    "Implicit detach timer");
   }
-  struct ue_mm_context_s* ue_context_p =
-      mme_app_get_ue_context_for_timer(mme_ue_s1ap_id, "Implicit detach timer");
+
   if (ue_context_p == NULL) {
     OAILOG_ERROR(
         LOG_MME_APP,
@@ -2040,6 +2053,7 @@ status_code_e mme_app_handle_initial_context_setup_rsp_timer_expiry(
   mme_ue_s1ap_id_t mme_ue_s1ap_id = 0;
   struct ue_mm_context_s* ue_context_p = NULL;
   if (timer_id == MME_APP_TIMER_INACTIVE_ID) {
+    // Expiry handler was called as part of timer recovery on service restart
     ue_context_p = (struct ue_mm_context_s*)args;
   } else {
     if (!mme_pop_timer_arg_ue_id(timer_id, &mme_ue_s1ap_id)) {
@@ -2437,14 +2451,19 @@ status_code_e mme_app_handle_paging_timer_expiry(zloop_t* loop, int timer_id,
                                                  void* args) {
   OAILOG_FUNC_IN(LOG_MME_APP);
   mme_ue_s1ap_id_t mme_ue_s1ap_id = 0;
-  if (!mme_pop_timer_arg_ue_id(timer_id, &mme_ue_s1ap_id)) {
-    OAILOG_WARNING(LOG_MME_APP, "Invalid Timer Id expiration, Timer Id: %u\n",
-                   timer_id);
-    OAILOG_FUNC_RETURN(LOG_MME_APP, RETURNok);
+  struct ue_mm_context_s* ue_context_p = NULL;
+  if (timer_id == MME_APP_TIMER_INACTIVE_ID) {
+    // Expiry handler was called as part of timer recovery on service restart
+    ue_context_p = (struct ue_mm_context_s*)args;
+  } else {
+    if (!mme_pop_timer_arg_ue_id(timer_id, &mme_ue_s1ap_id)) {
+      OAILOG_WARNING(LOG_MME_APP, "Invalid Timer Id expiration, Timer Id: %u\n",
+                     timer_id);
+      OAILOG_FUNC_RETURN(LOG_MME_APP, RETURNok);
+    }
+    ue_context_p =
+        mme_app_get_ue_context_for_timer(mme_ue_s1ap_id, "Paging timer");
   }
-  struct ue_mm_context_s* ue_context_p =
-      mme_app_get_ue_context_for_timer(mme_ue_s1ap_id, "Paging timer");
-
   if (ue_context_p == NULL) {
     OAILOG_ERROR(
         LOG_MME_APP,

--- a/lte/gateway/c/core/oai/tasks/mme_app/mme_app_bearer.c
+++ b/lte/gateway/c/core/oai/tasks/mme_app/mme_app_bearer.c
@@ -2038,13 +2038,19 @@ status_code_e mme_app_handle_initial_context_setup_rsp_timer_expiry(
     zloop_t* loop, int timer_id, void* args) {
   OAILOG_FUNC_IN(LOG_MME_APP);
   mme_ue_s1ap_id_t mme_ue_s1ap_id = 0;
-  if (!mme_pop_timer_arg_ue_id(timer_id, &mme_ue_s1ap_id)) {
-    OAILOG_ERROR(LOG_MME_APP, "Invalid Timer Id expiration, Timer Id: %u\n",
-                 timer_id);
-    OAILOG_FUNC_RETURN(LOG_MME_APP, RETURNok);
+  struct ue_mm_context_s* ue_context_p = NULL;
+  if (timer_id == MME_APP_TIMER_INACTIVE_ID) {
+    ue_context_p = (struct ue_mm_context_s*)args;
+  } else {
+    if (!mme_pop_timer_arg_ue_id(timer_id, &mme_ue_s1ap_id)) {
+      OAILOG_NOTICE(LOG_MME_APP, "Invalid Timer Id expiration, Timer Id: %u\n",
+                    timer_id);
+      OAILOG_FUNC_RETURN(LOG_MME_APP, RETURNok);
+    }
+    ue_context_p = mme_app_get_ue_context_for_timer(
+        mme_ue_s1ap_id, "Initial context setup response timer");
   }
-  struct ue_mm_context_s* ue_context_p = mme_app_get_ue_context_for_timer(
-      mme_ue_s1ap_id, "Initial context setup response timer");
+
   if (ue_context_p == NULL) {
     OAILOG_ERROR(
         LOG_MME_APP,

--- a/lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.c
+++ b/lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.c
@@ -914,6 +914,7 @@ void mme_ue_context_update_ue_sig_connection_state(
         MME_APP_TIMER_INACTIVE_ID) {
       mme_app_stop_timer(ue_context_p->mobile_reachability_timer.id);
       ue_context_p->mobile_reachability_timer.id = MME_APP_TIMER_INACTIVE_ID;
+      ue_context_p->time_mobile_reachability_timer_started = 0;
     }
     // Stop Implicit detach timer,if running
     if (ue_context_p->implicit_detach_timer.id != MME_APP_TIMER_INACTIVE_ID) {

--- a/lte/gateway/c/core/oai/tasks/mme_app/mme_app_timer_management.cpp
+++ b/lte/gateway/c/core/oai/tasks/mme_app/mme_app_timer_management.cpp
@@ -59,7 +59,10 @@ void mme_app_resume_timer(struct ue_mm_context_s* const ue_mm_context_pP,
    * from restart, so MME shall handle as timer expiry
    */
   if (timer->msec <= elapsed_time_in_ms) {
-    timer_expiry_handler(mme_app_task_zmq_ctx.event_loop, timer->id, NULL);
+    // In this case, the timer has no id since it is just getting resumed, pass
+    // the UE context to expiry handler
+    timer_expiry_handler(mme_app_task_zmq_ctx.event_loop,
+                         MME_APP_TIMER_INACTIVE_ID, (void*)ue_mm_context_pP);
     OAILOG_FUNC_OUT(LOG_MME_APP);
   }
   uint32_t remaining_time_in_msecs = timer->msec - elapsed_time_in_ms;

--- a/lte/gateway/c/core/oai/tasks/mme_app/mme_app_timer_management.cpp
+++ b/lte/gateway/c/core/oai/tasks/mme_app/mme_app_timer_management.cpp
@@ -61,8 +61,8 @@ void mme_app_resume_timer(struct ue_mm_context_s* const ue_mm_context_pP,
   if (timer->msec <= elapsed_time_in_ms) {
     // In this case, the timer has no id since it is just getting resumed, pass
     // the UE context to expiry handler
-    timer_expiry_handler(mme_app_task_zmq_ctx.event_loop,
-                         MME_APP_TIMER_INACTIVE_ID, (void*)ue_mm_context_pP);
+    timer_expiry_handler(mme_app_task_zmq_ctx.event_loop, timer->id,
+                         reinterpret_cast<void*>(ue_mm_context_pP));
     OAILOG_FUNC_OUT(LOG_MME_APP);
   }
   uint32_t remaining_time_in_msecs = timer->msec - elapsed_time_in_ms;

--- a/lte/gateway/python/integ_tests/s1aptests/test_ics_timer_expiry_with_mme_restart.py
+++ b/lte/gateway/python/integ_tests/s1aptests/test_ics_timer_expiry_with_mme_restart.py
@@ -115,6 +115,7 @@ class TestIcsTimerExpiryWithMmeRestart(unittest.TestCase):
             print("Waiting for", j, "seconds")
             time.sleep(1)
 
+        print("************************* Waiting for response from MME")
         response = self._s1ap_wrapper.s1_util.get_response()
         self.assertEqual(
             response.msg_type, s1ap_types.tfwCmd.UE_CTX_REL_IND.value,

